File: epi/Ruby/list_16.rb
=begin
Compute the median of a sorted circular linked list
Write a function that takes a pointer to an arbitrary node in a sorted circular
linked list, and returns the median of the linked list
=end
require "libraries"

def median(head)
  return nil if head.nil? || head.next.nil?
  largest = nil
  count = 0
  cur = head.next
  loop do
    count += 1
    largest = cur if cur.data > cur.next.data
    cur = cur.next
    break if cur == head.next
  end
  # now largest contains the biggest number in the sequence
  median = largest
  puts "largest: " + largest.to_s
  puts "count: " + count.to_s
  (count/2.0).ceil.times do
    median = median.next
  end
  if count % 2 == 1
    puts "count is odd"
    return median.data
  else
    puts "count is even"
    return (median.data + median.next.data) / 2.0
  end
end

b = Node.new(1, nil)
a = Node.new(2, Node.new(3, Node.new(4, Node.new(5, b))))
b.next = a
head = Node.new(nil, a)
puts median(head)

File: epi/Python/list_16.py
# compute the median of a sorted circular linked list

class Node:
  def __init__(self, data, node):
    self.data = data
    self.next = node

  def __str__(self):
    return str(self.data)

  def linkedlist(self):
    runner = self
    res = ''
    # The trick the replace do-for loop in python. I like it :D
    while True:
      runner = runner.next
      res += str(runner) + ' '
      if runner.next == self.next:
        break
    return res.strip()

def median(head):
  if head == None or head.next == None or head.next.next == None:
    return
  else:
    biggest = head.next # biggest will point to the smallest
    count = 0
    runner = head
    while True:
      runner = runner.next
      count += 1
      if biggest.data <= runner.data:
        biggest = runner
      if runner.next == head.next:
        break
  print 'biggest: ' + str(biggest)
  mid = biggest
  for dummy_i in range((count+1)/2):
    mid = mid.next
  print mid
  return mid.data if count&1 else (mid.data + mid.next.data)/2.0

if __name__ == '__main__':
  node6 = Node(6, None)
  head = Node(None, Node(17, Node(1, Node(2, Node(3, Node(4, Node(5, node6)))))))
  node6.next = head.next
  print head.linkedlist()
  print median(head)
  print '0 is true' if 0 else '0 is false'
